### PR TITLE
MM-55966 - Update ArrayFromJSON to use LimitedReader

### PR DESCRIPTION
**Pull request by sbishel**
#### Summary
This fixes a general set of issues where most Arrays passed to our RestAPIs are unbounded. Depending on the API and how the received arrays are used, the results of this may differ. 

This PR fixes this issue by 
 - creating a configuration setting for limiting the size of an accepted payload
 - using a LimitedReader when decoding the Payload body via ArrayFromJSON()
 - deduplicates the strings

Deduplicating Strings -
There is an existing RemoveDuplicateStrings function, which does an in-place removal and is performant. It does this by sorting the array. During testing, it was discovered that in some APIs the order of the strings matter. Therefore, another function (RemoveDuplicateStringNonSort)  was written to preserve the array order (via Map). It is not as performant and uses more memory, so it should only be used when preserving order is necessary.

Missing Unit Tests for updated functions - (created tickets under https://mattermost.atlassian.net/browse/MM-56030
channel_category.go (1)
dataretention.go (4)
status.go (1)
system.go (1)
user.go (1)

#### Ticket Link
Fixes https://mattermost.atlassian.net/browse/MM-55966

#### Release Note

```release-note
Adds a Limit to the payload size of API endpoints passing in Arrays
```

### CODE REVIEW

File: server/channels/api4/emoji_test.go
Comment (sbishel) [inline server/channels/api4/emoji_test.go:346]: This is the only unit test that required updating. The unit test depended on the order of returns, but the API call really shouldn't, so the unit test was updated.

**Review by sbishel — COMMENTED** [review]
Because of the call to RemoveDuplicateStrings(), most API returns will be now be returned sorted rather than the order they were requested. 

A couple places this is an issue was found via unit tests. Are there others? I attempted to detemine if this may cause a problem somewhere, but worry there may be something I missed.

File: server/public/model/utils.go
Comment (sbishel) [inline server/public/model/utils.go:484]: Debated about returning error, rather than empty array (as current implementation does). If we do, then all the utility functions should be updated. Thoughts?

File: server/public/model/config.go
Comment (agnivade) [inline server/public/model/config.go:401]: It's always a good practice to append the unit.

File: server/public/model/config.go
Comment (agnivade) [inline server/public/model/config.go:906]: Can we get a reasonable default from security? Was this arbitrarily chosen or was there a method to it?

File: server/public/model/utils.go
Comment (agnivade) [reply server/public/model/utils.go:484]: It's always good to return the error and bubble it up. We have an opportunity to do the right thing here. I'd say we should take it. Unfortunately, this means updating all those functions :(

File: server/public/model/utils_test.go
Comment (agnivade) [inline server/public/model/utils_test.go:220]: A slightly uncommon way to write a loop but okay :smile:

File: server/channels/api4/channel.go
Comment (hmhealey) [inline server/channels/api4/channel.go:428]: I wonder if we might want to force a limit reader for all API handlers by default by modifying `r.Body` before we even get here. In `ServeHTTP`, we could do something like
```go
r.Body = io.LimitReader(r.Body, maximumPayloadSize)
```
to have this apply anywhere. Maybe it could hold a reference to the original `r.Body` somehow so that the file upload APIs could get around this, but it would mean that we only need to make one change there and it'd apply automatically to all new APIs down the line

Comment (sbishel) [reply server/channels/api4/channel.go:428]: @hmhealey - Thanks for suggestion. Updating in `ServeHTTP` was discussed in [this document](https://docs.google.com/document/d/1v8nhBM10ZMsfzO8yovTCu3fcZiYt5prQLBSpYPnmjUg/edit). I'm willing to revisit, as we want to get this correct. 

We should also discuss the breaking change aspect. Is that ok for a security issue?

Comment (hmhealey) [reply server/channels/api4/channel.go:428]: Thanks for the link. I think doing it from `ServeHTTP` would make it a bit less likely that we accidentally re-introduce this issue down the line, but I don't think makes a huge difference either way.

As for the breaking change, I'm 0/5 on that since I don't know much about the Go ecosystem there. I just saw that something in the public package was changing, so I called it out. If we do care about it, we could just deprecate the current version and give your new ones a new name

File: server/public/model/config.go
Comment (sbishel) [reply server/public/model/config.go:906]: This value was essentially arbitrary but I have gone through several of the APIs, including ones that pass in an object looking for reasons this number might be limiting. I think this number is probably a bit higher even than necessary but too small risks breaking things. Also the reason for it being configurable.

@esarafianou - Do you have recommendations?

File: server/channels/api4/channel.go
Comment (agnivade) [reply server/channels/api4/channel.go]: @hmhealey - We discussed this. There are pros and cons to both approaches. Separating the file upload limit with the input body limit will require a separate middleware and a property to indicate which limit should be respected. We wanted to have a layered approach here - where at a high level, the file upload limit is respected. But different handlers are free to choose their own limit depending on the application logic.

To not re-introduce this again, we would be relying on vet checks to ensure people don't forget. 

There's more discussion on the google docs.

File: server/public/model/config.go
Comment (esarafianou) [reply server/public/model/config.go]: > I think this number is probably a bit higher even than necessary but too small risks breaking things. 

Agreed with that. I don't have any other recommendation.

**Comment by sbishel** [comment]
Ready for final review...I did create a new function and deprecate ArrayFromJSON().

File: server/channels/api4/channel.go
Comment (agnivade) [inline server/channels/api4/channel.go:430]: This actually gobbles up the returned error - making it to be no different than L441. Wondering if we can just convert the error to an AppError and set `c.Err = appError` and return. This way it's clear to the end user and also in the server logs as to what went wrong.

File: server/public/model/utils_test.go
Comment (agnivade) [inline server/public/model/utils_test.go:228]: Wondering if we can also verify the error returned here? From what I can see in the docs, it's `io.EOF`. So something like `require.True(t, errors.Is(err, io.EOF))` should work.

**Review by agnivade — CHANGES_REQUESTED** [review]
Just one comment on returning the underlying error.

File: server/public/model/utils_test.go
Comment (sbishel) [reply server/public/model/utils_test.go:229]: done

File: server/channels/api4/channel.go
Comment (sbishel) [reply server/channels/api4/channel.go]: Updated to return AppError in all cases.

**Review by agnivade — APPROVED** [review]
Thanks Scott!

**Comment by sbishel** [comment]
@esarafianou @hmhealey @streamer45 @mgdelacroix - Gentle ping to review this week if possible. This PR is necessary for rest of the payload issues to get fixed. Thanks.

**Comment by esarafianou** [comment]
@sbishel I'm waiting for the dev reviews to progress before I start the security one.

**Review by mgdelacroix — APPROVED** [review]
LGTM! :+1:

Left some nonblocking nits around `testify` helper improvements

File: server/channels/api4/user_local.go
Comment (streamer45) [inline server/channels/api4/user_local.go:254]: Small nit: `parseErr` would be my preference here naming wise rather than numbering.

File: server/channels/api4/channel.go
Comment (streamer45) [inline server/channels/api4/channel.go:711]: Should we check if `channelIds` is empty?

File: server/channels/api4/data_retention.go
Comment (streamer45) [inline server/channels/api4/data_retention.go:267]: Should we check if `teamIDs` is empty?

File: server/channels/api4/data_retention.go
Comment (streamer45) [inline server/channels/api4/data_retention.go:294]: Same as above.

File: server/channels/api4/data_retention.go
Comment (streamer45) [inline server/channels/api4/data_retention.go:386]: Same as above.

File: server/channels/api4/system.go
Comment (streamer45) [inline server/channels/api4/system.go:970]: Same as above.

File: server/channels/api4/channel.go
Comment (streamer45) [inline server/channels/api4/channel.go:909]: This could be problematic if length is zero the error could be nil. Also we are handling the empty case below already.

File: server/channels/api4/channel_category.go
Comment (streamer45) [inline server/channels/api4/channel_category.go:121]: For consistency I'd probably add an empty check here as well.

File: server/public/model/utils.go
Comment (streamer45) [inline server/public/model/utils.go:511]: Personally I'd prefer if the default case wasn't sorting vs the negated alternative. And then have a `SortedArrayFromJSON` version. It may also be worth reviewing where we actually need to sort as a potential performance improvement (can be deferred of course).

**Review by streamer45 — COMMENTED** [review]
Looks good overall, gave it a first pass and left mostly minor comments.

**Review by sbishel — COMMENTED** [review]
We need to decide how to dedup by default (sorted/unsorted) and if we are ok with slight functionality changes. (ie. error on no values).
I was trying to be safe when it comes to making too large of changes regarding functionality and performance, but if all agree, I will make the changes suggested.

File: server/channels/api4/channel.go
Comment (streamer45) [reply server/channels/api4/channel.go:711]: Understood. I'd expect a sane client not to send empty arrays but maybe we can defer this and plan a proper review since it may take more time to check everything and there are no reasons to delay the other improvements.

File: server/public/model/utils.go
Comment (streamer45) [reply server/public/model/utils.go:511]: Yeah, the explicit naming works for me. Not expecting to alter any functionality for the time being.

**Review by streamer45 — APPROVED** [review]
Thanks. Approving this to unblock the most important changes while we figure out the rest.

**Review by esarafianou — APPROVED** [review]
@sbishel Thanks for all the work on this one!

Just a question: if ArrayFromJSON is now deprecated, why are we not removing the function? Developers could end up keep using it out of habit

**Comment by sbishel** [comment]
> if ArrayFromJSON is now deprecated, why are we not removing the function? Developers could end up keep using it out of habit

I debated this one as well, because it is in a public library, we should give other potential consumers (plug-ins, etc) a chance to change it, rather than breaking immediately upon upgrade. We can add it to the [Vet check](https://mattermost.atlassian.net/browse/MM-55985) if necessary or I'll remove it. I'm not sure what or if we have a policy regarding deprecating in public library. I'm 0/5.

**Comment by agnivade** [comment]
FYI, this is one of the major motivations to split out to a separate module. Our new module version is still at 0.x series, meaning we can do whatever with it. I'll leave it up to you, but probably we would be fine to remove it entirely in the next ESR.

**Comment by sbishel** [comment]
I have created https://mattermost.atlassian.net/browse/MM-56363 to remove deprecated function...
@esarafianou - Did you QA as well?

**Comment by esarafianou** [comment]
@sbishel no I didn't QA it. Proper QA will probably need to check all the endpoints. If the QA folks can't take it, I can QA it but that will probably be the first week of January.

**Comment by sbishel** [comment]
Yeah. It will be the first week of January for anyone, Lindy is out as well. No worries, we'll get it in the new year.

**Comment by lindy65** [comment]
> @sbishel no I didn't QA it. Proper QA will probably need to check all the endpoints. If the QA folks can't take it, I can QA it but that will probably be the first week of January.

Hi @esarafianou - please will you QA this PR for us?  Many thanks! 🤗

**Comment by esarafianou** [comment]
@sbishel , @lindy65 It's been QAed. I didn't test all endpoints since they are reusing the same functions but tested some of them for both the sorted and the unsorted Array to Json.

**Comment by lindy65** [comment]
Thanks so much @esarafianou - much appreciated :)

**Comment by cwarnermm** [comment]
@sbishel - Does this PR have product documentation impact? If so, what details do our admins or end users need to know?

**Comment by sbishel** [comment]
@cwarnermm - No, this should be all "under the covers", no user impact.

**Pull request merged**
